l2geth/rollup/fees/rollup_fee.go

```diff
@@ -20,84 +24,143 @@ var (
-// the L2 gas limit will not be able to be decoded.
-// An explicit design goal of this scheme was to make the L2 gas limit be human
-// readable. The entire number is interpreted as the gas limit when computing
-// the fee, so increasing the L2 Gas limit will increase the fee paid.
-// The calculation is:
-// l1GasLimit = zero_count(data) * 4 + non_zero_count(data) * 16 + overhead
-// roundedL2GasLimit = ceilmod(l2GasLimit, 10_000)
-// l1Fee = l1GasPrice * l1GasLimit
-// l2Fee = l2GasPrice * roundedL2GasLimit
-// sum = l1Fee + l2Fee
-// scaled = sum / scalar
-// rounded = ceilmod(scaled, tenThousand)
-// roundedScaledL2GasLimit = roundedL2GasLimit / tenThousand
-// result = rounded + roundedScaledL2GasLimit
-// Note that for simplicity purposes, only the calldata is passed into this
-// function when in reality the RLP encoded transaction should be. The
-// additional cost is added to the overhead constant to prevent the need to RLP
-// encode transactions during calls to `eth_estimateGas`
-func EncodeTxGasLimit(data []byte, l1GasPrice, l2GasLimit, l2GasPrice *big.Int) *big.Int {
-	l1GasLimit := calculateL1GasLimit(data, overhead)
-	roundedL2GasLimit := Ceilmod(l2GasLimit, BigTenThousand)
-	l1Fee := new(big.Int).Mul(l1GasPrice, l1GasLimit)
-	l2Fee := new(big.Int).Mul(l2GasPrice, roundedL2GasLimit)
-	sum := new(big.Int).Add(l1Fee, l2Fee)
-	scaled := new(big.Int).Div(sum, bigFeeScalar)
-	rounded := Ceilmod(scaled, BigTenThousand)
-	roundedScaledL2GasLimit := new(big.Int).Div(roundedL2GasLimit, BigTenThousand)
-	result := new(big.Int).Add(rounded, roundedScaledL2GasLimit)
-	return result
+type Message interface {
+	From() common.Address
+	To() *common.Address
+	GasPrice() *big.Int
+	Gas() uint64
+	Value() *big.Int
+	Nonce() uint64
+	Data() []byte
 }
 
-func Ceilmod(a, b *big.Int) *big.Int {
-	remainder := new(big.Int).Mod(a, b)
-	if remainder.Cmp(common.Big0) == 0 {
-		return a
+type StateDb interface {
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+type RollupOracle interface {
+	SuggestL1GasPrice(ctx context.Context) (*big.Int, error)
+	SuggestL2GasPrice(ctx context.Context) (*big.Int, error)
+	SuggestOverhead(ctx context.Context) (*big.Int, error)
+	SuggestScalar(ctx context.Context) (*big.Int, error)
+}
+
+func CalculateFee(tx *types.Transaction, gpo RollupOracle) (*big.Int, error) {
+	// Read the variables from the cache
+	l1GasPrice, err := gpo.SuggestL1GasPrice(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	overhead, err := gpo.SuggestOverhead(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	scalar, err := gpo.SuggestScalar(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := rawTransaction(tx, false)
+	if err != nil {
+		return nil, err
+	}
+
+	l1Fee := CalculateL1Fee(raw, overhead, l1GasPrice, scalar)
+	l2GasLimit := new(big.Int).SetUint64(tx.Gas())
+	l2Fee := new(big.Int).Mul(tx.GasPrice(), l2GasLimit)
```

Comment:
I think this function needs a comment explaining what it does. It's a bit confusing since there are other similar function names.